### ISSUE TITLE
quaternion_exp_to_log returns NaN in float16 for any zero-vector-part quaternion (identity included): the default eps=1e-8 underflows to 0

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3966]
### Describe the bug

`quaternion_exp_to_log` returns `NaN` in `float16` for **every** quaternion whose vector part is zero — including the identity quaternion `(1, 0, 0, 0)`, whose log is unambiguously the origin. `float32`, `float64` and `bfloat16` all return `[0, 0, 0]` for the same inputs.

The cause is a dtype-dependent underflow of the default `eps`, not the quaternion maths. The implementation guards the division with

```python
norm_q = torch.norm(quaternion_vector, p=2, dim=-1, keepdim=True).clamp(min=eps)
```

with `eps: float = 1e-8` (`kornia/geometry/conversions.py`, `quaternion_exp_to_log`). `float16`'s smallest subnormal is ≈ 6e-8, so `1e-8` rounds to exactly `0.0` and the clamp becomes a no-op. `acos(±1)` is then multiplied by a zero vector part and divided by a zero norm: `0 / 0 → NaN`.

`bfloat16` is unaffected because its exponent range is float32's, so `1e-8` is representable there.

### Reproduction steps

```python
import torch
from kornia.geometry.conversions import quaternion_exp_to_log

for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=dtype)   # identity quaternion
    print(dtype, quaternion_exp_to_log(q).tolist())
```

```
torch.float64  [0.0, 0.0, 0.0]
torch.float32  [0.0, 0.0, 0.0]
torch.bfloat16 [0.0, 0.0, 0.0]
torch.float16  [nan, nan, nan]
```

It is not specific to the identity — any zero vector part reproduces it, in both halves of the double cover:

```python
q = torch.tensor([-1.0, 0.0, 0.0, 0.0], dtype=torch.float16)
quaternion_exp_to_log(q)          # tensor([nan, nan, nan], dtype=torch.float16)
```

The eps underflow is the whole story — passing an eps that survives `float16` fixes it:

```python
q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=torch.float16)
quaternion_exp_to_log(q, eps=1e-3)          # tensor([0., 0., 0.], dtype=torch.float16)
float(torch.tensor(1e-8, dtype=torch.float16))   # 0.0      <- the default underflows
float(torch.tensor(1e-8, dtype=torch.bfloat16))  # 1.0011717677116394e-08
```

A non-zero vector part is fine at the same dtype, which is why the suite has never caught it:

```python
quaternion_exp_to_log(torch.tensor([-1.0, 1e-3, 0.0, 0.0], dtype=torch.float16))
# tensor([3.1406, 0.0000, 0.0000], dtype=torch.float16)
```

### Expected behavior

`quaternion_exp_to_log` should return the origin `[0, 0, 0]` for a quaternion with zero vector part at every floating dtype it accepts, as it already does at `float64`, `float32` and `bfloat16`. Whatever the fix, a float-typed default that silently rounds to zero in a supported dtype is a trap: candidates are a dtype-aware epsilon (`torch.finfo(q.dtype).tiny` or `eps`), clamping the norm after casting, or selecting the origin explicitly where the vector part is zero.

### Environment

- kornia 0.9.0rc1 (`main` at 8454a81a), torch 2.9.1, CPU, macOS (arm64). Reproduced on MPS as well.

### Notes

- Found while auditing the rotation-representation conversions for per-op Convention blocks (batch 3b of #3941). It is the same failure *class* as the float16 eps underflow already pinned for the pixel-coordinate normalizers in #3942 (`test_wart_float16_underflowed_default_eps_flips_branches`) and adjacent to #3939 — a float default epsilon that is not representable in every supported dtype. A repo-wide sweep for `clamp(min=eps)` with a float default under float16 is probably worth doing once, rather than one issue per site.
- Distinct from #3953 (`quaternion_exp_to_log` silently accepting non-unit input) and from #3949 (NaN *gradients* at the identity in `quaternion_to_axis_angle` / `axis_angle_to_quaternion`): this is a NaN in the **forward** output, at a valid unit quaternion, in float16 only.

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by ducha-aiki** [comment]
Follow-up: this is a **family**, not a single site. While finishing the batch-3b convention blocks I found the same float16 epsilon underflow in two more functions in the same module, one of which is the root cause of a NaN reported separately in #3952.

Every one of these is a Python `float` default that is not representable in float16 (smallest subnormal ≈ `5.960464477539063e-08`), so it rounds to `0.0` and the `clamp(min=eps)` guard it feeds becomes a no-op:

| function | default `eps` | float16 representation | float16 symptom |
|---|---|---|---|
| `quaternion_exp_to_log` | `1e-8` | `0.0` | NaN for any zero-vector-part quaternion, incl. the identity (this issue) |
| `quaternion_log_to_exp` | `1e-8` | `0.0` | `quaternion_log_to_exp(zeros(3))` → `[1.0, nan, nan, nan]` |
| `normalize_quaternion` | `1e-12` | `0.0` | `nq(zeros(4))` and `nq([1e-13,0,0,0])` → all-NaN; propagates, so `quaternion_to_rotation_matrix(zeros(4))` is an all-NaN 3×3 at float16 while it is the identity at float32 (#3952) |

```python
import torch
from kornia.geometry.conversions import quaternion_log_to_exp as l2e, normalize_quaternion as nq

l2e(torch.zeros(3, dtype=torch.float16))            # tensor([1., nan, nan, nan], dtype=torch.float16)
l2e(torch.zeros(3, dtype=torch.float16), eps=1e-3)  # tensor([1., 0., 0., 0.], dtype=torch.float16)
nq(torch.zeros(4, dtype=torch.float16))             # tensor([nan, nan, nan, nan], dtype=torch.float16)
```

bfloat16 escapes all three, because its exponent range is float32's and `1e-8` is representable there.

The shape of a fix that covers the family: derive the epsilon from the input dtype (`torch.finfo(x.dtype).tiny`, or `eps` scaled to the dtype) instead of a module-level float constant, at every site that clamps a norm. Fixing them one at a time will keep producing this bug, since the constant is copied between functions.

I have not swept the rest of the library — a `grep` for `clamp(min=eps)` and similar guards with float defaults, checked at float16, would say whether this extends past `geometry/conversions.py`. Flagging rather than filing that, since it is a maintainer call how wide to scope one issue.

Posted on behalf of @ducha-aiki by Claude (Fable 5).
